### PR TITLE
Sysinfo: support configuration files installed system wide [v2]

### PR DESCRIPTION
**Pull request by clebergnu**
Commit 768455c3 removed the existing configuration files, and with
that, the patching that happened on the RPM SPEC file changing the
sysinfo defaults to the system wide /etc based locations.

Without a configuration file to tweak, we're left wither with patching
code, or making the code more dynamic and respecting the system wide
configuration paths no matter if it's an RPM based installation or
not.  This chooses the later.

Fixes: https://github.com/avocado-framework/avocado/issues/4208
Signed-off-by: Cleber Rosa <crosa@redhat.com>

---

Changes from v1 (#4508):
* Added docstring (@pevogam)
* Properly handle absolute paths given as input (@pevogam)

### CODE REVIEW

**Review by willianrampazzo — APPROVED** [review]

**Review by beraldoleal — APPROVED** [review]
LGTM

**Comment by beraldoleal** [comment]
@clebergnu since this is a LTS fix, would this be ported to 82lts too?

**Comment by clebergnu** [comment]
> @clebergnu since this is a LTS fix, would this be ported to 82lts too?

Correct.  Once merged, I'll send a backport to 82LTS.

**Review by richtja — APPROVED** [review]

File: avocado/core/utils.py
Comment (pevogam) [inline avocado/core/utils.py:40]: Could this be
```python
"""
Returns either a system wide path, or one relative to the base.

...
"""
```
according to the pydocstyle which was used in the nrunner (e.g. Runnable class) module for example?

**Review by pevogam — APPROVED** [review]
Just finished testing and this fixes the situation. Awesome!

Also left one final small comment if you don't mind. Other than that, LGTM.

**Comment by pevogam** [comment]
@beraldoleal I thought you would wait for me here.

**Comment by pevogam** [comment]
Honestly, it is a bit offensive considering I have reported this issue and waited for months to see it resolved and then you couldn't wait even for a day, especially seeing that I was pinged and the comments above.

**Comment by beraldoleal** [comment]
@pevogam sorry, it was my bad, I only saw your comments right after I merged.

Since your two comments on v1 were incorporated here, I assumed that you would be ok with a merge here.

I will personally handle your docstring suggestion if @clebergnu is ok with the change.

**Comment by pevogam** [comment]
> @pevogam sorry, it was my bad, I only saw your comments right after I merged.

It is a good maintainer habit to first read the history and see if a PR is ok to merge from all participants. As I am just a contributor I cannot mark the PR as waiting for review. And as we take the the v1-v2-vN approach of multiple pushes hiding the PR's history (something I have already complained about), my previous review gets hidden. If you would like to keep the v1-vN approach of multiple PR-s at least for now you would have to account for this drawback too (not mentioned in my original list of drawbacks perhaps) and check previous versions. At least I see no way in which I could aid you and compensate for this.

> Since your two comments on v1 were incorporated here, I assumed that you would be ok with a merge here.

I would be ok but the fact that my test passed here could be just a lucky draw. There might have also been the outcome that they would fail and detect breakage in which case this would be prematurely merged. So the lucky outcome doesn't justify the actions.

**Comment by beraldoleal** [comment]
> > @pevogam sorry, it was my bad, I only saw your comments right after I merged.
> 
> It is a good maintainer habit to first read the history and see if a PR is ok to merge from all participants but indeed, it is more complicated in this case because of the current workflow. As I am just a contributor I cannot mark the PR as waiting for review. And as we take the the v1-v2-vN approach of multiple pushes hiding the PR's history (something I have already complained about), my previous review gets hidden. If you would like to keep the v1-vN approach of multiple PR-s at least for now you would have to account for this drawback too (not mentioned in my original list of drawbacks perhaps) and check previous versions. At least I see no way in which I could aid you and compensate for this.

Yes, indeed. I could not see your previous comments here, since v1 was closed. And IMO this is completely fine. When a PR is closed in favor of a new one, we can assume that the contributor has discussed all the topics and will incorporate the suggestions. We don't have to review all versions, it is impractical for some cases where we can reach a huge number of interactions, and sometimes, the v9 is a completely different approach from the first version.

Although it is a good practice to wait for reporters and confirm if the issue was solved, afaict, we don't have any policy to enforce that. We have users that report issues and disappear (not your case). Because of that, I'm not in favor of enforcing this, and I don't think this should be a blocker as well.

> > Since your two comments on v1 were incorporated here, I assumed that you would be ok with a merge here.
> 
> I would be ok but the fact that my test passed here could be just a lucky draw. There might have also been the outcome that they would fail and detect breakage in which case this would be prematurely merged. So the lucky outcome doesn't justify the actions.

This is why we have the code review process. We had on this PR 3 approvals, and tests were passing on all our  CI checks. According to our Contributors Guide, my merge was not premature. Our current requirements are:

```
Every single Pull Request in Avocado has to be reviewed by at least one other developer. All members of the core team have permission to merge a Pull Request, but there are some conditions that have to be fulfilled before merging the code:

    1. Pull Request has to pass the CI tests.
    2. One ‘Approved’ code review should be given.
    3. No explicit disapproval should be present.

```

Yes, I totally agree that we could improve this, adding more reviewers and giving more time before merges. But this is another discussion. 

I might have a different view from others, but IMO master branch is a development branch. While we do a huge effort to not break things, we are humans and unfortunately, things could break. Reverts and amend PRs could be done easily before releases and LTS.

So, said that, I'm totally in favor of discussing new ways to improve our review process, like I said we have room for improvements.
My only suggestion is to start a new discussion topic to get better visibility.

**Comment by pevogam** [comment]
> > > @pevogam sorry, it was my bad, I only saw your comments right after I merged.
> > 
> > 
> > It is a good maintainer habit to first read the history and see if a PR is ok to merge from all participants but indeed, it is more complicated in this case because of the current workflow. As I am just a contributor I cannot mark the PR as waiting for review. And as we take the the v1-v2-vN approach of multiple pushes hiding the PR's history (something I have already complained about), my previous review gets hidden. If you would like to keep the v1-vN approach of multiple PR-s at least for now you would have to account for this drawback too (not mentioned in my original list of drawbacks perhaps) and check previous versions. At least I see no way in which I could aid you and compensate for this.
> 
> Yes, indeed. I could not see your previous comments here, since v1 was closed. And IMO this is completely fine. When a PR is closed in favor of a new one, we can assume that the contributor has discussed all the topics and will incorporate the suggestions. We don't have to review all versions, it is impractical for some cases where we can reach a huge number of interactions, and sometimes, the v9 is a completely different approach from the first version.

I don't agree it is fine. There are cases where the contributor might not incorporate the exact suggestions or misunderstand what the reviewer meant. This is why a reviewer typically checks the new version again with further comments. I cannot speak about cases where the PR review back and forth becomes very long but even then there are people involved and TODOs alnog the way that have to be kept track of. And a normal PR would on GitHub would keep this history.

> Although it is a good practice to wait for reporters and confirm if the issue was solved, afaict, we don't have any policy to enforce that. We have users that report issues and disappear (not your case). Because of that, I'm not in favor of enforcing this, and I don't think this should be a blocker as well.

So you think disregarding reviewers the way it happened here using the current workflow is fine then? Because I don't see an alternative solution here. I actually find it ironical that I started the discussion against the multiple GitHub pushes and the same issue came back to bite me now. Even if contributors don't write back which indeed could happen, I think this should happen after a timeout of at least a week and not just a day later the way it happened here.

> > > Since your two comments on v1 were incorporated here, I assumed that you would be ok with a merge here.
> > 
> > 
> > I would be ok but the fact that my test passed here could be just a lucky draw. There might have also been the outcome that they would fail and detect breakage in which case this would be prematurely merged. So the lucky outcome doesn't justify the actions.
> 
> This is why we have the code review process. We had on this PR 3 approvals, and tests were passing on all our CI checks. According to our Contributors Guide, my merge was not premature. Our current requirements are:
> 
> ```
> Every single Pull Request in Avocado has to be reviewed by at least one other developer. All members of the core team have permission to merge a Pull Request, but there are some conditions that have to be fulfilled before merging the code:
> 
>     1. Pull Request has to pass the CI tests.
>     2. One ‘Approved’ code review should be given.
>     3. No explicit disapproval should be present.
> ```

You are right, according to this guideline it is not a violation. Then I guess in order to adjust to these requirements I should always disapprove first and do so for each new PR version you push (also never forget to disapprove) in order to make sure some of the discussed plans during the PR (like asking somebody to test) are complied with.

> Yes, I totally agree that we could improve this, adding more reviewers and giving more time before merges. But this is another discussion.
> 
> I might have a different view from others, but IMO master branch is a development branch. While we do a huge effort to not break things, we are humans and unfortunately, things could break. Reverts and amend PRs could be done easily before releases and LTS.

We are at least paritally on keeping-master-stable boat considering we do automated checks and CI before each merge though. But in any case, not breaking master wasn't my point here. Simply that we cannot use a lucky circumstance to conclude the workflow is fine.

> So, said that, I'm totally in favor of discussing new ways to improve our review process, like I said we have room for improvements.
> My only suggestion is to start a new discussion topic to get better visibility.

Yes agreed. Feel free to start one if you want. My main pain point here was and remains the multiple PR versions, something for which there is already an ongoing discussion.

**Comment by pevogam** [comment]
> > ```
> > Every single Pull Request in Avocado has to be reviewed by at least one other developer. All members of the core team have permission to merge a Pull Request, but there are some conditions that have to be fulfilled before merging the code:
> > 
> >     1. Pull Request has to pass the CI tests.
> >     2. One ‘Approved’ code review should be given.
> >     3. No explicit disapproval should be present.
> > ```
> 
> 
> You are right, according to this guideline it is not a violation. Then I guess in order to adjust to these requirements I should always disapprove first and do so for each new PR version that is pushed (also never forget to disapprove) in order to make sure some of the discussed plans during the PR (like asking somebody to test) are complied with.

There might not be an easy solution here indeed but if the PR page is kept the same one could always see who are the current reviewers and if somebody is still being waited on. With new PR-s previous reviewers disappear and one has to search their previous comments again to refresh what needed to be changed while in parallel re-registering themselves as somebody participating in the PR.

**Comment by beraldoleal** [comment]
> > Although it is a good practice to wait for reporters and confirm if the issue was solved, afaict, we don't have any policy to enforce that. We have users that report issues and disappear (not your case). Because of that, I'm not in favor of enforcing this, and I don't think this should be a blocker as well.
> 
> So you think disregarding reviewers the way it happened here using the current workflow is fine then? Because I don't see an alternative solution here. I actually find it ironical that I started the discussion against the multiple GitHub pushes and the same issue came back to bite me now. Even if contributors don't write back which indeed could happen, I think this should happen after a timeout of at least a week and not just a day later the way it happened here.

Of course not, my first comment was saying that I would take care personally of your concerns. But I also don't agree that reviewers/maintainers should look to the entire version history to find gaps. This is impractical as I said, even with a single PR. Discussions could be a long journey. From my side, please always assume good intentions. Is not because our review process isn't perfect that we are being disrespectful.

The solution is to improve our review process, giving more time to reviewers to manifest their disagreement. Maybe one week is too much, but I agree.

> > This is why we have the code review process. We had on this PR 3 approvals, and tests were passing on all our CI checks. According to our Contributors Guide, my merge was not premature. Our current requirements are:
> > ```
> > Every single Pull Request in Avocado has to be reviewed by at least one other developer. All members of the core team have permission to merge a Pull Request, but there are some conditions that have to be fulfilled before merging the code:
> > 
> >     1. Pull Request has to pass the CI tests.
> >     2. One ‘Approved’ code review should be given.
> >     3. No explicit disapproval should be present.
> > ```
> 
> You are right, according to this guideline it is not a violation. Then I guess in order to adjust to these requirements I should always disapprove first and do so for each new PR version that is pushed (also never forget to disapprove) in order to make sure some of the discussed plans during the PR (like asking somebody to test) are complied with.

Well, if you disapprove with good reasons then is fine. But if you are doing this just because you disagree with our contributor guide, then this might not be very welcome.

One thing that we are doing when we would like to review a PR, is we mark ourselves as "reviewers" (even it is not documented). I understand that you might be missing some permissions right now, but we can investigate better our current permissions and roles and see how to fix this situation. 

IMO, this is much more polite than starting disapproving new PRs. 

> > Yes, I totally agree that we could improve this, adding more reviewers and giving more time before merges. But this is another discussion.
> > I might have a different view from others, but IMO master branch is a development branch. While we do a huge effort to not break things, we are humans and unfortunately, things could break. Reverts and amend PRs could be done easily before releases and LTS.
> 
> We are at least partially on keeping-master-stable boat considering we do automated checks and CI before each merge though. But in any case, not breaking master wasn't my point here. Simply that we cannot use a lucky circumstance to conclude the workflow is fine.

Unfortunately, we have today 69% of coverage (and we are working to improve this). But having CI checks before the merges, doesn't give us "lucky-free" guarantees, *today*. Most of our issues are detected by users. That is the reason we have intermediate releases and an LTS release. 

Again, I understand your complaint but please always try to assume good intentions here.

File: avocado/core/utils.py
Comment (clebergnu) [reply avocado/core/utils.py:40]: Actually it looks like we were *not* following https://www.python.org/dev/peps/pep-0257/ in this regard, so I'm trying to use this new style whenever I write new docstrings.

**Comment by clebergnu** [comment]
> Honestly, it is a bit offensive considering I have reported this issue and waited for months to see it resolved and then you couldn't wait even for a day, especially seeing that I was pinged and the comments above (in v1 but this is your choice of workflow) where @clebergnu mentioned he would wait. I explicitly mentioned I need just a day.

Hi @pevogam . I hope you understand that it was a mistake, and the intention was never to shut you out of the review process.  I can vouch for @beraldoleal in the sense that he is *very* concerned with having a welcoming environment for all in the project.

Now, I hope we can make a lemonade out of these bitter lemons.  Like I mentioned before, I will be spending time next sprint to review the contribution guidelines/workflow, so that it gets substantially improved.

In the mean time, it looks like this could have been prevented (not the only way though) if you could have been assigned as a reviewer.  Would you mind if I set you as a contributor/maintainer/whatever permission needed to be assigned for reviews?

Thanks for testing it, and accept my sincere apologies for your experience.

**Comment by pevogam** [comment]
Hi @beraldoleal @clebergnu, I have no doubt that this is a sincere mistake and sorry if I sounded too harsh about it, giving you the impression that I seek intentions behind this. My point here is only to help improve the process and this situation illustrated another drawback of the current approach.

> IMO, this is much more polite than starting disapproving new PRs.

For sure my point here wasn't to seriously disapprove like this but exactly to illustrate the absurdity of the situation. But being practical, I one idea that I have and doesn't require admin privileges is to request changes explaining in a comment that I am testing the given branch. This is perhaps a middle ground that doesn't require too much authorization (and you are the proper maintainers after all) and could signal Beraldo or somebody else about to merge that there is a blocker. Of course I don't intend to do this and disappear for long or anything like this, I usually try to respond within days and on a few occasions I was quite busy and could react within a week or two.

> One thing that we are doing when we would like to review a PR, is we mark ourselves as "reviewers" (even it is not documented). I understand that you might be missing some permissions right now, but we can investigate better our current permissions and roles and see how to fix this situation.

Yes, this is the natural way but perhaps the idea above is simpler to implement - all we have to do is for me and other contributors to follow such a workflow to deal with the multi-PR# pushes.

> Now, I hope we can make a lemonade out of these bitter lemons. Like I mentioned before, I will be spending time next sprint to review the contribution guidelines/workflow, so that it gets substantially improved.

No problem, let's see where the future brings us considering some more usage and the current plans.

Thank you both for the understanding! As the PR is already merged there isn't much to discuss here any longer, I am sure we will figure it out with more interaction.

**Pull request merged**
